### DISCUSSION TITLE
Dedupe takes longer time and all of a sudden it fails

### DISCUSSION THREAD

**Discussion by DevanandamS2013** [moj-analytical-services/splink#159] [Q&A]
- I tried performing Dedupe with my data which has 5 million records(column names: Id, name, postal_code, country).

- In my blocking rule I used(**l.postal_code = r.postal_code and l.country = r.country(together**)) to reduce as much as record I can.

- Based on the comparison score on **column:"name"** I should able to link data)

- The issue here is , the process is really slow(very slow) and in log I can able to see "not enough memory to cache **rdd xxx"(log attached below).**

- I can able to find reason behind it. In my data I have count of **postal_code: "23xxx" as 65k** and with many other postal_code as well(count is really high). (attached below)

- So while performing join between these postal code , it couldnt able to cache the rdd completely (I guess).

**What I tried:**
As instructed by your guide on working with large data,
-I played around with **spark.sql.shuffle.partitions** and **spark.default.parallelism**
-I couldn't able to do further to tighten my blocking rule(so its not in scope)
-I tried with **"break the lineage"** as well but it takes long time to process.
-I tried increasing my cluster size as well.

Could you please propose a solution to with work this data? like what should be the required Cluster size and what modifications has to be done in my configuration to make it run.

**_Postal_code count:_**

![image](https://user-images.githubusercontent.com/49081926/102541990-6f8f8d80-40b1-11eb-9db4-3ce2cf0bcd1e.png)

**_log:_**

![image](https://user-images.githubusercontent.com/49081926/102542031-7a4a2280-40b1-11eb-9328-1553e0efb417.png)

**Comment by AnasAlzahrani** [comment]
I'm using a local machine with 35g memory. Also I'm doing big data linkage with millions of records. 
This is my conf set:
```
conf = (conf.setMaster('local[*]')
        .set('spark.executor.memory', '4G')
        .set('spark.driver.memory', '22G')
        .set('spark.driver.maxResultSize', '29G')
        .set("spark.sql.shuffle.partitions", "1000")
        .set("spark.default.parallelism", "1000"))
sc = SparkContext(conf=conf)
sc.setCheckpointDir("temp_graphframes/")
```

and I persist on my disk only

```
from pyspark import StorageLevel

df_l.persist(StorageLevel.DISK_ONLY)
df_r.persist(StorageLevel.DISK_ONLY)
df_e.persist(StorageLevel.DISK_ONLY)
```

include this code (as recommended in the [demo](https://github.com/moj-analytical-services/splink_demos/blob/master/large_jobs_and_breaking_lineage.ipynb) 

```
def blocked_comparisons_to_disk(df, spark):
    df.write.mode("overwrite").parquet("gammas_temp/")
    df_new = spark.read.parquet("gammas_temp/")
    return df_new

def scored_comparisons_to_disk(df, spark):
    df.write.mode("overwrite").parquet("df_e_temp/")
    df_new = spark.read.parquet("df_e_temp/")
    return df_new
```

Then 

```
from splink import Splink

linker = Splink(settings, spark, 
                df_l=df_l, 
                df_r=df_r, 
                break_lineage_blocked_comparisons = blocked_comparisons_to_disk,
                break_lineage_scored_comparisons = scored_comparisons_to_disk)

df_e = linker.get_scored_comparisons()
```

In my setting, I set "retain_intermediate_calculation_columns": False and sometime "max_iterations": 0 or 3

I hope this is useful for you. If you come with any idea please share it with us. That would be nice as I'm straggling with memory problem too!

**Comment by RobinL** [comment]
@DevanandamS2013 If the only information you have is postcode and country, and you have some postcodes which appear >1 million times, then i think you probably simply don't have enough information to dedupe.  

When you block on postcode, it will compare all rows within a postcode to all other rows within that postcode - the exact formula for the number of rows generated for each postcode (which appears n times) is is n*(n-1)/2.    As a result, you generally need to keep n in the hundreds to prevent extremely large numbers of comparisons being generated.

**Reply by DevanandamS2013** [reply]
> @DevanandamS2013 If the only information you have is postcode and country, and you have some postcodes which appear >1 million times, then i think you probably simply don't have enough information to dedupe.
> 
> When you block on postcode, it will compare all rows within a postcode to all other rows within that postcode - the exactly formula is n*(n-1)/2. As a result, you generally need to keep n in the hundreds to prevent extremely large numbers of comparisons being generated.

@RobinL Thanks for the explanation. Yeah, you are right. I will try to block more records as proposed by @mamonu

**Comment by mamonu** [comment]
I propose to perhaps block a bit more?
Usually something that still can allow a lot of the true matches to pass the blocking pass would be to check that the first characters of the surnames are the same.  So in your case you would need to create 2 derived columns with firstname and surname out of the name column.

In the blocking pass it helps to block the obvious mismatches so the algorithm has the chance to work on candicate pairs that have
some chance of being similar. 

Usually names are mispelled/misheard towards the last few letters so there is some logic behind checking the first letter (and this common prefix thing is quite important as Winkler specifically reasoned about it and created the Jaro-Winkler improvement on the Jaro comparator)

**Reply by DevanandamS2013** [reply]
@mamonu Thanks for the idea. I'll give a try. :)
